### ISSUE TITLE
How to retrieve validation results after training

### ISSUE THREAD

**Issue by haimat** [ultralytics/ultralytics#855]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

When I train a model via

```python
model = YOLO("yolov8m.pt")
model.train(...)
```

then in the end the validation results are printed to the console like this:

```bash
Validating training/ski-defects-sample/weights/best.pt...
Ultralytics YOLOv8.0.26 🚀 Python-3.10.6 torch-1.13.0+cu116 CUDA:0 (NVIDIA GeForce RTX 2070 with Max-Q Design, 7982MiB)
Model summary (fused): 218 layers, 25840339 parameters, 0 gradients, 78.7 GFLOPs
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:01<00:00,  1.11it/s]
                   all         25         64      0.892      0.907      0.963      0.736
Speed: 1.1ms pre-process, 9.3ms inference, 0.0ms loss, 0.6ms post-process per image
Results saved to training/ski-defects-sample
```

This is great, now I would need exactly those validation metrics (P, R, etc.) printed to the console.
How can I retrieve these validation metrics from the call to `train()`?

**Comment by 2016bgeyer** [comment]
results = model.train(...)
results = model.val()  # evaluate model performance on the validation set

**Comment by Laughing-q** [comment]
@haimat please refer to our docs [https://docs.ultralytics.com/tasks/detection/#val](https://docs.ultralytics.com/tasks/detection/#val) . :)
```python
metrics = model.val()  # no arguments needed, dataset and settings remembered
metrics.box.map    # map50-95
metrics.box.map50  # map50
metrics.box.map75  # map75
metrics.box.maps   # a list contains map50-95 of each category
metrics.box.mp    # P
metrics.box.mr    # R
```

**Comment by haimat** [comment]
@2016bgeyer @Laughing-q 
Hey guys, thanks for your response. I tried that already, and it didn't work. The return value of `model.val()` is `None` ☹️

**Comment by Laughing-q** [comment]
@haimat please upgrade the package and try again, we actually implemented this functionality yesterday. :)

**Comment by haimat** [comment]
Yes thanks, now it works as documented, thanks :+1:

**Comment by SV1367** [comment]
I have a related question:
How to get the validation results for each class?
I am trying yolo val ..... but I only get the "all" result

**Comment by Laughing-q** [comment]
@SV1367 
```python
metrics.box.maps   # a list contains map50-95 of each category
```

**Comment by sivikt** [comment]
Hey folks, 
is any way to store all these metrics in some readable format (besides serializing them as pickle)?

**Comment by glenn-jocher** [comment]
You can save the validation metrics to a JSON file by setting `save_json=True` in your validation command.

**Comment by sivikt** [comment]
well, yeah but i thought about saving the whole object in some readable format (images that it plots are not very convenient, not interactive) and results contain more info.. Pickling does not work either, since DetResults contains generators.

**Comment by glenn-jocher** [comment]
@sivikt you can save the metrics to a JSON file using `save_json=True` during validation, which provides a readable format for further analysis.
